core/drivers/versal_mbox.c

```diff
@@ -302,99 +428,109 @@ TEE_Result versal_mbox_alloc(size_t len, const void *init,
 	return TEE_SUCCESS;
 }
 
-TEE_Result versal_mbox_notify(struct versal_ipi_cmd *cmd,
+TEE_Result versal_mbox_free(struct versal_mbox_mem *mem)
+{
+	if (!mem)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	free(mem->buf);
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_mbox_notify(struct versal_ipi *ipi,
+			      struct versal_ipi_cmd *cmd,
 			      struct versal_ipi_cmd *rsp, uint32_t *err)
 {
 	TEE_Result ret = TEE_SUCCESS;
-	uint32_t remote_status = 0;
+	uint32_t status = 0;
 
-	mutex_lock(&ipi.lock);
+	mutex_lock(&ipi->lock);
 
-	ret = versal_mbox_write_req(cmd);
+	ret = versal_mbox_write_req(ipi, cmd);
 	if (ret) {
 		EMSG("Can't write the request command");
 		goto out;
 	}
 
-	if (IS_ENABLED(CFG_VERSAL_TRACE_MBOX))
-		versal_mbox_call_trace(cmd->data[0]);
+	/* Trigger interrupt to remote */
+	io_write32(ipi->regs + IPI_TRIG_OFFSET,
+		   IPI_BIT_MASK(ipi->rmt));
 
-	ret = mbox_call(IPI_MAILBOX_NOTIFY, IPI_BLOCK);
-	if (ret) {
-		EMSG("IPI error");
-		goto out;
-	}
+	/* Wait for remote to acknowledge the interrupt */
+	do {
+		status = io_read32(ipi->regs + IPI_OBR_OFFSET);
+	} while (status & IPI_BIT_MASK(ipi->rmt));
```

Comment:
timeout?  (maybe see `IO_READ32_POLL_TIMEOUT()`)

Reply:
@jcorbier a timeout would be good here. If the remote processor hangs and never responds then this will hang the mailbox driver/OP-TEE.

Reply:
Should be addressed by fixup commit [6d01a34](https://github.com/OP-TEE/optee_os/pull/6738/commits/6d01a341403ac150ba780df976a9c4903c3f435a)